### PR TITLE
Enable package node migration for Dynamo Core 2.0+ graphs

### PR DESCRIPTION
**Pull request by alfarok**
### Purpose

[QNTM-5547](https://jira.autodesk.com/browse/QNTM-5547)

This PR restores the previous (1.3 and older) behavior for package node migration for JSON graphs (2.0+).  For example, you can now migrate a packaged node that was built against Core 2.0 but was later renamed.  We previously handled 1.3 (XML) to 2.0 (JSON) but never 2.0 to 2.0+.

**Package Migration Strategies:**
ZeroTouch nodes => Use a `Migrations.XML` file located in the packages `bin` folder
NodeModel derived nodes => Use the `AlsoKnownAs` attribute on the class

**Examples:**

`MyZeroTouchLib.MyNodes.SayHello` to `MyZeroTouchLib.MyNodes.SayHelloRENAMED`
```XML
<?xml version="1.0"?>
<migrations>
  <priorNameHint>
    <oldName>MyZeroTouchLib.MyNodes.SayHello</oldName>
    <newName>MyZeroTouchLib.MyNodes.SayHelloRENAMED</newName>
  </priorNameHint>
</migrations>
```

`SampleLibraryUI.Examples.DropDownExample` to `SampleLibraryUI.Examples.DropDownExampleRENAMED`
```C#
namespace SampleLibraryUI.Examples
{
    [NodeName("Drop Down Example")]
    [NodeDescription("An example drop down node.")]
    [IsDesignScriptCompatible]
    [AlsoKnownAs("SampleLibraryUI.Examples.DropDownExample")]
    public class DropDownExampleRENAMED : DSDropDownBase
    {
        ...
    }
{
```

### Testing
1) Created a test package with several node types (not committed)
2) Created a legacy graph including these nodes (committed)
3) Renamed the classes of all included nodes
4) Included `migrations.xml` for ZeroTouch nodes
5) Added `[AlsoKnownAs]` attribute for the NodeModel node
6) Rebuilt package (committed)
7) Assert all nodes migrate to the appropriate types and no dummy nodes are found

Test Cases:
- [x] ZeroTouch single input (overload)
- [x] ZeroTouch multi input (overload)
- [x] ZeroTouch no input multi output
- [x] NodeModel dropdown

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner @aparajit-pratap @QilongTang 

### FYIs

@johnpierson

### CODE REVIEW

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 1]
```diff
@@ -169,7 +169,16 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
             else if (typeof(DSFunctionBase).IsAssignableFrom(type))
             {
                 var mangledName = obj["FunctionSignature"].Value<string>();
-                var functionDescriptor = libraryServices.GetFunctionDescriptor(mangledName);
+                var priorNames = libraryServices.GetPriorNames();
+                FunctionDescriptor functionDescriptor;
+
+                // Attempt to located a newer migrated version of the node
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: sic(locate)?

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 2]
```diff
@@ -169,7 +169,16 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
+                if (priorNames.ContainsKey(mangledName))
```
Comment (aparajit-pratap) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: Just use `TryGetValue `. You can check for the key and get the value at the key at the same time.

Comment (alfarok) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: I actually was using that originally but was hoping to figure out how to eliminate having to define `functionDescriptor` without assigning a value.  I ended up having it anyway so I agree and will switch back.  I also have a few other changes coming, this doesn't actually solve all the migration scenarios properly yet.

```C#
priorNames.TryGetValues(mangledName, out functionDescriptor)
```

**Comment by QilongTang** [comment]
@alfarok  Question on nodeModel nodes renaming, instead of dong what you put as example, can I rename the node in the opposite way below?
```
namespace SampleLibraryUI.Examples
{
    [NodeName("Drop Down Example")]
    [NodeDescription("An example drop down node.")]
    [IsDesignScriptCompatible]
    [AlsoKnownAs("SampleLibraryUI.Examples.DropDownExampleRENAMED ")]
    public class DropDownExample : DSDropDownBase
    {
        ...
    }
{
```

**Comment by mjkkirschner** [comment]
@alfarok did you ever investigate if you could get rid of this ```              else if (type == typeof(DSVarArgFunction))
             {
                 var functionId = Guid.Parse(obj["FunctionSignature"].Value<string>());
                 node = manager.CreateCustomNodeInstance(functionId);
             } ```?

**Comment by mjkkirschner** [comment]
I think some tests will break as they count the number of packages loaded from the test/pkgs folder.

**Comment by alfarok** [comment]
@QilongTang I haven't tried that but my assumption is that it would migrate the node in the opposite direction?  I think it makes more sense for the classes to be renamed so nodes can more clearly be traced back to their origin and keeps the 2 in sync.  The library basically does the opposite when it was reorganized and now none of the locations truly represent the code locations/classes.

**Comment by alfarok** [comment]
@mjkkirschner I don't think it is required but am also a bit nervous about removing it thinking it may have been added for some strange edge case?  What do you think?

Also the self-serve is running one last time so I will update any tests that rely of the package count if required.

**Comment by QilongTang** [comment]
@alfarok Good to know! I think people need to be aware of this is the preferred way

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 3]
```diff
@@ -56,6 +57,20 @@ public NodeReadConverter(CustomNodeManager manager, LibraryServices libraryServi
             }
         }
 
+        public NodeReadConverter(CustomNodeManager manager, LibraryServices libraryServices, NodeFactory nodeFactory, bool isTestMode = false)
+        {
+            this.manager = manager;
+            this.libraryServices = libraryServices;
+            this.nodeFactory = nodeFactory;
+            this.isTestMode = isTestMode;
```
Comment (aparajit-pratap) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: Minor but I think we avoid using `this` as per existing coding conventions (which may have been forgotten along the way). Also if you use Resharper you'll see that a lot of these things are highlighted by the tool.

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 4]
```diff
@@ -106,6 +121,18 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
+            // Check for and attempt to resolve an unknown type before proceeding
+            if (type == null)
+            {
+                // Check to see if the missing type can be resolved (AlsoKnownAs)
+                var unresolvedName = obj["$type"].Value<string>().Split(',').FirstOrDefault();
+                Type newType;
+                nodeFactory.ResolveType(unresolvedName, out newType);
+
+                if (newType != null) // If resolved update the type
+                { type = newType; }
```
Comment (aparajit-pratap) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: Here it is preferred to have a single curly brace on a line (_don't know what happened to coding conventions once followed by the team_):
```
// If resolved update the type
if (newType != null) 
{ 
type = newType; 
}
```

Comment (mjkkirschner) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards

**Comment by aparajit-pratap** [comment]
@alfarok other than the harping about coding conventions this looks good to me. Thanks!

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs (hunk 4 above)
Comment (QilongTang) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: oh wow, if we would like to get this topic up, please fix all the comments in this PR so they are in the same format 😉 
//[space][Comment]

Comment (alfarok) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: Sorry comments without a space were copied from the constructor I made obsolete, I would never commit such a crime 😂

**Comment by alfarok** [comment]
@aparajit-pratap 🚨 🚓 guilty as charged, I can't say I've ever seen that doc before though

**Comment by mjkkirschner** [comment]
LGTM
